testing/A_intruduction.md

```diff
@@ -0,0 +1,518 @@
+
+As we saw near the beginning of this guide, we can run our entire suite of tests with `mix test`.
+
+```console
+$ mix test
+....
+
+Finished in 0.2 seconds (0.1s on load, 0.03s on tests)
+4 tests, 0 failures
+
+Randomized with seed 540755
+```
+
+If we would like to run all the tests in a given directory, `test/controllers` for instance, we can pass the path to that directory to `mix test`.
+
+```console
+$ mix test test/controllers/
+.
+
+Finished in 0.2 seconds (0.1s on load, 0.04s on tests)
+1 tests, 0 failures
+
+Randomized with seed 652376
+```
+
+In order to run all the tests in a specific file, we can pass the path to that file into `mix test`.
+
+```console
+$ mix test test/views/error_view_test.exs
+...
+
+Finished in 0.2 seconds (0.2s on load, 0.00s on tests)
+3 tests, 0 failures
+
+Randomized with seed 220535
+```
+
+And we can run a single test in a file by appending a colon and a line number to the filename.
+
+Let's say we only wanted to run the test for the way `HelloPhoenix.ErrorView` renders `500.html. The test begins on line 12 of the file, so this is how we would do it.
+
+```console
+$ mix test test/views/error_view_test.exs:12
+Including tags: [line: "12"]
+Excluding tags: [:test]
+
+.
+
+Finished in 0.1 seconds (0.1s on load, 0.00s on tests)
+1 tests, 0 failures
+
+Randomized with seed 288117
+```
+
+We chose to run this specifying the first line of the test, but actually, any line of that test will do. These line numbers would all work - `:13`, `:14`, or `:15`.
+
+### Running Tests Using Tags
+
+ExUnit allows us to tag our tests at the case level or on the individual test level. We can then choose to run only the tests with a specific tag, or we can exclude tests with that tag and run everything else.
+
+Let's experiment with how this works.
+
+First, we'll add a `@moduletag` to `test/views/error_view_test.exs`.
+
+```elixir
+defmodule HelloPhoenix.ErrorViewTest do
+  use HelloPhoenix.ConnCase, async: true
+
+  @moduletag :error_view_case
+
+  . . .
+```
+
+If we use only an atom for our module tag, ExUnit assumes that it has a value of `true`. We could also specify a different value if we wanted.
+
+```elixir
+defmodule HelloPhoenix.ErrorViewTest do
+  use HelloPhoenix.ConnCase, async: true
+
+  @moduletag error_view_case: "some_interesting_value"
+
+  . . .
+```
+
+For now, let's leave it as a simple atom `@moduletag :error_view_case`.
+
+We can run only the tests from the error view case by passing `--only error_view_case` into `mix test`.
+
+```console
+$ mix test --only error_view_case
+Including tags: [:error_view_case]
+Excluding tags: [:test]
+
+...
+
+Finished in 0.1 seconds (0.1s on load, 0.00s on tests)
+3 tests, 0 failures
+
+Randomized with seed 125659
+```
+
+> Note: ExUnit tells us exactly which tags it is including and excluding for each test run. If we look back to the previous section on running tests, we'll see that line numbers specified for individual tests are actually treated as tags.
```

Comment:
Ah, I see the same format for other notes further in the guide so I'm guessing it's intentional :)